src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,704 @@
+    assert!(prev_sum.len() == num_saturating_sum_bits);
+
+    let mut carry_in = SB::ZERO;
+    let mut output = vec![];
+    for i in 0..num_saturating_sum_bits {
+        let c = ctx.narrow(&BitOpStep::from(i));
+        let (sum_bit, carry_out) = if i < num_trigger_value_bits {
+            one_bit_adder(c, record_id, &cur_value[i], &prev_sum[i], &carry_in).await?
+        } else {
+            one_bit_adder(c, record_id, &SB::ZERO, &prev_sum[i], &carry_in).await?
+        };
+
+        output.push(sum_bit);
+        carry_in = carry_out;
+    }
+    let updated_is_saturated = -carry_in
+        .clone()
+        .multiply(
+            prev_is_saturated,
+            ctx.narrow(&BitOpStep::from(num_saturating_sum_bits)),
+            record_id,
+        )
+        .await?
+        + &carry_in
+        + prev_is_saturated;
+    Ok((BitDecomposed::new(output), updated_is_saturated))
+}
+
+///
+/// Returns (`difference_bit`, `carry_out`)
+///
+async fn one_bit_subtractor<C, SB>(
+    ctx: C,
+    record_id: RecordId,
+    x: &SB,
+    y: &SB,
+    carry_in: &SB,
+) -> Result<(SB, SB), Error>
+where
+    C: UpgradedContext<Gf2, Share = SB>,
+    SB: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
+{
+    // compute difference bit as not_y XOR x XOR carry_in
+    let difference_bit = SB::share_known_value(&ctx, Gf2::ONE) - y + x + carry_in;
+
+    let x_xor_carry_in = x.clone() + carry_in;
+    let y_xor_carry_in = y.clone() + carry_in;
+    let not_y_xor_carry_in = SB::share_known_value(&ctx, Gf2::ONE) - &y_xor_carry_in;
+
+    let carry_out = x_xor_carry_in
+        .multiply(&not_y_xor_carry_in, ctx, record_id)
+        .await?
+        + carry_in;
+
+    Ok((difference_bit, carry_out))
+}
+
+///
+/// TODO: optimize this
+/// We can avoid doing this many multiplications given the foreknowledge that we are always subtracting from zero
+/// There's also no reason to compute the `carry_out` for the final bit since it will go unused
+///
+async fn compute_truncated_difference_to_cap<C, SB>(
+    ctx: C,
+    record_id: RecordId,
+    cur_sum: &BitDecomposed<SB>,
+    num_trigger_value_bits: usize,
+    num_saturating_sum_bits: usize,
+) -> Result<BitDecomposed<SB>, Error>
+where
+    C: UpgradedContext<Gf2, Share = SB>,
+    SB: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
+{
+    assert!(cur_sum.len() == num_saturating_sum_bits);
+
+    let mut carry_in = SB::share_known_value(&ctx, Gf2::ONE);
+    let mut output = vec![];
+    for (i, bit) in cur_sum.iter().enumerate().take(num_trigger_value_bits) {
+        let c = ctx.narrow(&BitOpStep::from(i));
+        let (difference_bit, carry_out) =
+            one_bit_subtractor(c, record_id, &SB::ZERO, bit, &carry_in).await?;
+
+        output.push(difference_bit);
+        carry_in = carry_out;
+    }
+    Ok(BitDecomposed::new(output))
+}
+
+#[allow(clippy::too_many_lines, clippy::too_many_arguments)]
+async fn compute_row_with_previous<C, BK, TV>(
```

Comment:
I know that we have an excel file that does this calculation, but it would be nice to put a paragraph or two, describing the algorithm at the very high level

Reply:
Yes :). I'll do that. Also the excel file is not public - so a comment is necessary.